plugin/plugin.lua

```diff
@@ -55,88 +60,91 @@ local function initialize_configuration(config)
-        return nil
+   
+    local cors_allow_headers = {}
+    local requested_headers = ngx.req.get_headers()['access-control-request-headers']
+    if requested_headers then
+        for header in requested_headers:gmatch('[^,%s]+') do
+            table.insert(cors_allow_headers, header)
+        end
     end
 
-    return encryption_key_bytes
+    return cors_allow_headers
 end
 
+--
+-- Most CORS response headers must be returned during OPTIONS requests
+-- A subset of headers are needed during the main request
+--
 local function add_cors_response_headers(config, is_error)
 
     local origin = ngx.req.get_headers()['origin']
     if origin and array_has_value(config.trusted_web_origins, origin) then
-
+        
         if config.cors_enabled or is_error then
-
-            -- For plugin errors we always add these CORS headers, so that the SPA can read the error response body
             ngx.header['access-control-allow-origin'] = origin
             ngx.header['access-control-allow-credentials'] = 'true'
-            if #config.trusted_web_origins > 1 then
-                ngx.header['vary'] = 'origin'
-            end
         end
 
         if config.cors_enabled then
 
+            local vary = 'origin'
             local method = ngx.req.get_method():upper()
             if method == 'OPTIONS' then
-                if config.cors_allow_methods then
+
+                if #config.cors_allow_methods > 0 then
                     local allow_methods_str = table.concat(config.cors_allow_methods, ',')
-                    if allow_methods_str then
-                        ngx.header['access-control-allow-methods'] = allow_methods_str
-                    end
+                    ngx.header['access-control-allow-methods'] = allow_methods_str
                 end
-            end
 
-            if config.cors_allow_headers then
-                local allow_headers_str = table.concat(config.cors_allow_headers, ',')
-                if allow_headers_str then
+                local cors_allow_headers = get_cors_allowed_headers(config)
+                if #cors_allow_headers > 0 then
+                    local allow_headers_str = table.concat(cors_allow_headers, ',')
                     ngx.header['access-control-allow-headers'] = allow_headers_str
                 end
-            end
-
-            if config.cors_expose_headers then
-                local expose_headers_str = table.concat(config.cors_expose_headers, ',')
-                if expose_headers_str then
-                    ngx.header['access-control-expose-headers'] = expose_headers_str
-                end
-            end
-            
-            if config.cors_max_age then
+                
                 if config.cors_max_age > 0 then
                     ngx.header['access-control-max-age'] = config.cors_max_age
                 end
+
+                vary = 'origin,access-control-request-headers'
```

Comment:
Should `access-control-request-headers` always be present in the `Vary` header? Or should there be a check whether it was present in the request?